### ISSUE TITLE
Memory requirements

### ISSUE THREAD

**Issue by grofte** [sayakpaul/Sharpness-Aware-Minimization-TensorFlow#9]
Hi,

Great work. From the notebook it looks like the runtime is about a third longer with SAM. Is that right? That's not too bad. 

But how much more memory does the training require? We are calculating gradients twice for each step right? Don't we need more memory for that?

PS If I set the learning rate to 1e-2 the model without SAM quite outperforms the model with SAM  
*With*  
`Epoch 97/100
49/49 [==============================] - 4s 78ms/step - loss: 0.5391 - accuracy: 0.7984 - val_loss: 0.5719 - val_accuracy: 0.8091 - lr: 0.0025`  
*Without*  
`Epoch 53/200
49/49 [==============================] - 5s 101ms/step - loss: 0.2032 - accuracy: 0.9299 - val_loss: 0.4490 - val_accuracy: 0.8669 - lr: 7.8125e-05`

**Comment by sayakpaul** [comment]
Yeah, it's supposed to be less efficient than the standard ERM training. I admittedly didn't try out the hyperparameters that would showcase SAM's prowess as my main focus was to implement its workflow in a readable manner. Sorry if the README didn't clarify that. I did list out the differences in my implementation [here](https://github.com/sayakpaul/Sharpness-Aware-Minimization-TensorFlow#notes).

On a similar note, you could check out [this work (ESAM)](https://openreview.net/forum?id=n0OeTdNRG0Q) that addresses SAM's inefficiency with respect to training time and memory requirements.
